js/src/Zone.js

```diff
@@ -0,0 +1,91 @@
+import ExpectedMark from "./ExpectedMark";
+import UXCapture from "./UXCapture";
+
+/**
+ * Zone represents collection of elements groupped together and corresponding phase of page load
+ *
+ * Only one set of zones can be tracked for the same view at one point in time
+ */
+export default class Zone {
+  // constructs individual zone object
+  constructor(config) {
+    // {
+    //   name: "ux-destination-verified",
+    //   marks: ["ux-image-online-logo", "ux-image-inline-logo"]
+    //   selectors: ["img.logo"],
+    //   startMark: "navigationStart",
+    //   onMeasure: measureName => {}
+    //   onMark: markName => {}
+    // }
+
+    // name to be used for UserTiming measures
+    this.measureName = config.name;
+
+    // callback to execute when Zone is complete
+    this.onMeasure = config.onMeasure;
+
+    // callback for marks to call when they are complete
+    this.onMark = config.onMark;
+
+    // array of mark names that were recorded so far
+    this.recordedMarkNames = [];
+
+    this.startMarkName = config.startMarkName || "navigationStart";
+
+    // look up existing mark object or create a new one
+
+    // now just string names for UserTiming marks
+    // in the future, different types of events, e.g. ImageElement("logo"), PaintTimer("first-paint")
+    this.marks = config.marks.map(markName => {
+      const mark = ExpectedMark.create(markName);
+
+      mark.onComplete(mark => {
+        // if Zone's onMark callback is specified, call it with mark name
+        if (this.onMark) {
+          this.onMark(mark.name);
+        }
+
+        if (this.checkCompletion(mark)) {
+          this.complete(mark);
+        }
+      });
+
+      return mark;
+    });
+  }
+
+  /**
+   * Callback to be called by each mark in the zone upon recording
+   *
+   * @param {Mark} recordedMark latest mark recorded
+   */
+  checkCompletion(recordedMark) {
```

Comment:
This is a little confusing to me - 
This method does two things, 1. it marks the latest `recordedMark` as 'complete' and 2. it checks the completion of the other `recordedMarkNames` right?

Nitpick - I would expect `checkCompletion` to only do the second thing.
Questions - Why is `recordedMarkNames` an array?

Reply:
Good catch, it's supposed to be an object if we want to use key uniqueness here, you know what, I should probably rewrite this to check UserTiming API instead of internal function.

Reply:
Rewritten.

Reply:
👍